docs/hugo/content/design/ADR-2022-09-Multiple-Credential-Operator.md

```diff
@@ -0,0 +1,153 @@
+
+## Options
+
+In all the proposed solutions, we would need to cache the credentials to be used for resources when we reconcile and a mechanism to identify the credentials used for a resource. 
+
+### Option 1: Secrets lookup
+
+In secrets lookup option, we would need a fixed secret pattern like below. Which then will be looked up for each time a new resource is created. Operator would look up for the secret pattern for resource group first, then namespace. If secret exists, cache the credentials for the resource to be used for the next time or use the default global.
+
+NOTE: Annotation/condition will be added on the resource while creation which will have information about the secret name used for creation/reconcile.
+
+**Fixed secret name pattern:**
+```
+Namespaced secret -> aso-credential (in the namespace)
+Per-resource-group secret -> aso-credential-{rg-name}
+Per-resource secret -> aso-credential-{resource-name}
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly
+
+**Cons:**
+1. Not very flexible, as users don't have much control for which resource they want to use which credential
+
+### Option 2: Configuration using annotations
+
+In Configuration using annotations option, we'll use annotations like below on the resource to fetch credentials from a secret and cache it. Will default to global if secret with name not found or no annotation provided.
+
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Flexible, as users have the control for which resource they want to use which credential
+
+**Cons:**
+1. Will have to specify annotation on each resource explicitly
+
+### Option 3: Secret lookup + configuration using annotations
+
+Here, we would use a fixed pattern for namespaces secrets + annotations for per-resource-group and per-resource secrets. Operator would look up for the annotation(`serviceoperator.azure.com/credential-from`) on resource created, if annotation found, then use credentials from secret in annotation. If not, then look for the namespaced secret and use the namespaced credential.If none provided, operator shall continue with the global credential.
+
+NOTE: If namespaced secret or global secret is used, operator would inject `serviceoperator.azure.com/credential-from` annotation with the secret name value. 
+
+**Fixed namespaced secret name pattern:**
+```
+Namespaced secret -> aso-credential (in the namespace)
+```
+
+**Annotation for per-resource-group and per-resource secret**
+```
+annotations: 
+  serviceoperator.azure.com/credential-from: any-secret
+```
+
+**Pros:**
+1. Users don't have to specify the configuration for each resource explicitly if using a namespaced secret
+2. Users get the flexibility of over-riding the credential they want to use as per resource and as per resource group
+
+**Cons:**
+1. Could be complicated/confusing for users to understand the usage and credentials being used
```

Comment:
You could call out here the mitigation mentioned above of automatically adding the `credential-from` annotation when they use at least the namespace scope secret mode, so that it's more clear to them which secret is being used.

I personally think that mitigation makes this problem pretty minimal. Curious what @theunrepentantgeek thinks.